### PR TITLE
[SPARK-22838][Core] Avoid unnecessary copying of data

### PR DESCRIPTION
**Pull request by ConeyLiu**
## What changes were proposed in this pull request?

If we read data from FileChannel to HeapByteBuffer, there is a need to copy the data from the off-heap to the on-heap, you can see the follow code:

```java
static int read(FileDescriptor var0, ByteBuffer var1, long var2, NativeDispatcher var4) throws IOException {
    if(var1.isReadOnly()) {
      throw new IllegalArgumentException("Read-only buffer");
    } else if(var1 instanceof DirectBuffer) {
      return readIntoNativeBuffer(var0, var1, var2, var4);
    } else {
      ByteBuffer var5 = Util.getTemporaryDirectBuffer(var1.remaining());

      int var7;
      try {
        int var6 = readIntoNativeBuffer(var0, var5, var2, var4);
        var5.flip();
        if(var6 > 0) {
          var1.put(var5);
        }

        var7 = var6;
      } finally {
        Util.offerFirstTemporaryDirectBuffer(var5);
      }

      return var7;
    }
  }
```

## How was this patch tested?

Existing UT.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by ConeyLiu** [comment]
@cloud-fan Please take a look, thanks a lot.

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
cc @jiangxb1987

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 1]
```diff
@@ -208,7 +209,7 @@ private class EncryptedBlockData(
     conf: SparkConf,
     key: Array[Byte]) extends BlockData {
 
-  override def toInputStream(): InputStream = Channels.newInputStream(open())
+  override def toInputStream(): InputStream = new NioBufferedFileInputStream(file)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: will `NioBufferedFileInputStream` release its memory automatically?

Comment (ConeyLiu) [reply core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: You meaning the memory buffer? The `NioBufferedFileInputStream` has `close` method, you can see follow:
org.apache.spark.io.NioBufferedFileInputStream.java
```java
@Override
  public synchronized void close() throws IOException {
    fileChannel.close();
    StorageUtils.dispose(byteBuffer);
  }

  @Override
  protected void finalize() throws IOException {
    close();
  }
```

Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: This should create a `CryptoInputStream` right? Why could we simply new a `NioBufferedFileInputStream` instead? Am I missing something?

Comment (ConeyLiu) [reply core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: Sorry for the mistake, updated.

File: core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java [hunk 2]
```diff
@@ -61,6 +61,7 @@ private boolean refill() throws IOException {
         nRead = fileChannel.read(byteBuffer);
       }
       if (nRead < 0) {
+        byteBuffer.flip();
```
Comment (ConeyLiu) [inline core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java]: This should be a bug.  The byteBuffer is fiped after `fileChannel.read(byteBuffer)`, so if then we can read the remaining  dirty data. The test case will be add latter. I'm not sure this is should be another patch. @cloud-fan

File: core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java [hunk 3]
```diff
@@ -91,7 +92,12 @@ public synchronized int read(byte[] b, int offset, int len) throws IOException {
 
   @Override
   public synchronized int available() throws IOException {
-    return byteBuffer.remaining();
+    int n = byteBuffer.remaining();
+    long avail = fileChannel.size() - fileChannel.position();
+    long total = avail + n;
+    return avail > (Long.MAX_VALUE - n)
+      ? Integer.MAX_VALUE : total > Integer.MAX_VALUE
+      ? Integer.MAX_VALUE : (int)total;
```
Comment (ConeyLiu) [inline core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java]: This also shoule be fixed.

File: core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java (hunk 2 above)
Comment (jiangxb1987) [reply core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java]: Doesn't `nRead < 0` means we actually don't read any bytes?

Comment (ConeyLiu) [reply core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java]: This related to this error: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85228/testReport/org.apache.spark.storage/BlockManagerSuite/LRU_with_mixed_storage_levels_and_streams__encryption___on_/
I'm not very sure the reason, but I guess this happens such as follow:
```scala
var i = 0
while (i < inputStream.avaiable()) {
     //do something
}
```
After we arrived at the end of the file which `i == (inputStream.avaiable() - 1)`, then we get `-1` from `inputStream.read()`. And this time we need to call the `refill()` too. Even if we can't get the data from the underlying `fileChannel`, but the `byteBuffer` flipped. So the `inputStream.avaiable` changed, and we still can read the dirty data remained in the `byteBuffer`.

**Comment by ConeyLiu** [comment]
It seems the error not related. And can you add me to the whitelist?

**Comment by jiangxb1987** [comment]
Seems you should rebase with the latest master.

File: core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java (hunk 2 above)
Comment (jiangxb1987) [reply core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java]: Sorry I don't get your point, would you mind share the link to the code block you mention to? Thanks!

Comment (ConeyLiu) [reply core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java]: The problem is when we reach the end of file, we call  `NioBufferedFileInputStream.call()` twice can't get the `-1` both.
you can see here:
org.apache.commons.crypto.stream.input.StreamInput.java
```java
public int read(ByteBuffer dst) throws IOException {
        int remaining = dst.remaining();
        int read = 0;

        while(remaining > 0) {
            int n = this.in.read(this.buf, 0, Math.min(remaining, this.bufferSize));
            if (n == -1) {
                if (read == 0) {
                    read = -1;
                }
                break;
            }

            if (n > 0) {
                dst.put(this.buf, 0, n);
                read += n;
                remaining -= n;
            }
        }

        return read;
    }
```
1. reading the data from the underlying `NioBufferedFileInputStream`, read some remaining bytes.
2. reaching the end, so this time we call the `byteBuffer.flip()`. After `flip()`, the `byteBuffer.hasRemaining()` changed to true, because the `flip()` changed the limit and position.
3. But the return of `StreamInput.read()` is larger than 0 which read from step 1.
4. So we will call the `StreamInput.read()` . But the `byteBuffer.hasRemaining()` of `NioBufferedFileInputStream` changed, so we will read the dirty data from the `byteBuffer`.

I'm not very sure this explanation is correct enough.

**Comment by ConeyLiu** [comment]
I'll update it tomorrow.

**Comment by ConeyLiu** [comment]
cc @jiangxb1987 any comments on this?

**Comment by cloud-fan** [comment]
cc @jerryshao

File: core/src/main/scala/org/apache/spark/storage/DiskStore.scala [hunk 4]
```diff
@@ -152,7 +153,7 @@ private class DiskBlockData(
     file: File,
     blockSize: Long) extends BlockData {
 
-  override def toInputStream(): InputStream = new FileInputStream(file)
+  override def toInputStream(): InputStream = new NioBufferedFileInputStream(file)
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: IIUC, the returned `InputStream` will be deserialized in `BlockManger`, And deserializer will copy the data from direct memory to on-heap memory, otherwise how do we visit POJO?

So unless if we purely manipulate binary data, otherwise we have to copy the data to on-heap. Please correct me if I'm wrong.

Besides, I think this is not the hotspot, so memory copying should not bring in big overhead.

Comment (ConeyLiu) [reply core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: Hi @jerryshao, thanks for reviewing. This is inspired by #15408. 
> the returned `InputStream` will be deserialized in `BlockManger`

This is not entirely correct. Sometimes we don't need deserialized, such as network transmission. And also, this does not add extra work to deserialization, but reduces the effort of network-like delivery.

Comment (jerryshao) [reply core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: IIUC for network (netty) transmission, it uses zero copy sendFile, which is another path (`toNetty`). Here `toInputStream` seems only used by `BlockManager` from code.

Also this is not the critical path, reducing memory copy will not increase the performance a lot. Besides, due to the complexity of the code stack, we can guarantee there's no memory-copy in the whole code chain, so I doubt about the final improvements of changes.

Comment (ConeyLiu) [reply core/src/main/scala/org/apache/spark/storage/DiskStore.scala]: >IIUC for network (netty) transmission, it uses zero copy sendFile, which is another path (toNetty). 

Thanks for explaining, I did not notice this before.

**Comment by ConeyLiu** [comment]
close it, thanks for everyone.
